game-app/game-core/src/main/java/games/strategy/triplea/ui/UnitChooser.java

```diff
@@ -648,12 +683,183 @@ public Dimension getMinimumSize() {
+                  * (1 + category.getDependents().size());
+
+        if (spaceRequiredForNonWithdrawableIcon) {
+          width += uiContextDecoration.nonWithdrawableImage.getWidth(null) + PADDING;
+        }
+
+        final int height = uiContextDecoration.unitImageFactory.getUnitImageHeight();
         return new Dimension(width, height);
       }
+
+      @VisibleForTesting
+      public ChooserEntry getEmbeddingChooserEntry() {
+        return ChooserEntry.this;
+      }
     }
   }
+
+  /**
+   * once <code>UnitChooser</code> is divided up in a model and a presentation
+   *
+   * (With <code>ChooserEntry</code> being a candidate to go into the model
+   * and <code>UnitChooserEntryIcon</code> being a candidate to go into the presentation,
+   * the method <code>getImage</code> should probably move to <code>UnitChooserEntryIcon</code>.
+   *
+   * It can't be there for the moment, because <code>UnitChooserEntryIcon</code> is an
+   * inner class, and static methods are not supported in inner classes at language level 11.)
+   *
+   * @param category category of the image
+   * @param forceDamaged true iff the image should show a damaged unit
+   * @param uiContext context for loading the image resource
+   * @return the unit image, possibly decorated by the non-withdrawable-icon.
+   */
+
+  @VisibleForTesting
+  public static Image getImage(final UnitCategory category,
+                               final boolean forceDamaged,
+                               final UiContext uiContext) {
+    final var uiContextDecoration = UiContextDecoration.get(uiContext);
+    final Image image = uiContextDecoration.unitImageFactory
+        .getImage(
+            ImageKey.builder()
+                .type(category.getType())
+                .player(category.getOwner())
+                .damaged(forceDamaged || category.hasDamageOrBombingUnitDamage())
+                .disabled(category.getDisabled())
+                .build())
+        .orElseThrow();
+
+    return category.getCanRetreat()
+        ? image
+        : uiContextDecoration.getUnitImageWithNonWithdrawableImage(image);
+  }
+
+  static class UiContextDecoration {
+    private static UiContextDecoration latestUCD;
+
+    @VisibleForTesting
+    public final UiContext uiContext;
+    @VisibleForTesting
+    public final UnitImageFactory unitImageFactory;
+    @VisibleForTesting
+    public final BufferedImage nonWithdrawableImage;
+    private final Map<Image,BufferedImage> unitImageWithNonWithrawableImage =
+        Collections.synchronizedMap(new HashMap<>());
+
+    private UiContextDecoration(final UiContext uiContext) {
+      this.uiContext = uiContext;
+      final var unitImageFactory = uiContext.getUnitImageFactory();
+      final int unitImageHeight = unitImageFactory.getUnitImageHeight();
+      final double nonWithdrawableImageHeight = getNonWithdrawableImageHeight(unitImageHeight);
+      nonWithdrawableImage = loadNonWidthdrawableImage((int) nonWithdrawableImageHeight);
```

Comment:
This line always generates the 'nonWithdrawableImage' even if it is not rendered. Can we fix this so we load and generate the non-withdrawable image only when it is actually used?

If we have the code to get the non-withdrawable image in the 'image factory' entirely, then it could be easier to do.

It is getting weird though by using  'engine image loader' to get the raw images, but then we are using an image factory (which is a glorified wrapper around resource-loader) to get image height. It seems like either 'unit image factory' should be not at all involved here, or it should be more than 100% involved and contain all code for finding and generating the correct 'withdrawable icon'. In such a case it has a resource loader and we can use that to load the needed images (which makes changing 'EngineImageLoader' unnecessary).

Reply:
Will do that once the functionality is at the right place (`UnitImageFactory` probably)